Fix comment formatting and prevent potential XSS attack

ChrisLolz opened the pull request:
Hi, this is a pretty important issue so I hope this gets resolved as soon as possible.

While I was experimenting with the extension, I noticed that there was an edge case in your html sanitation where you could inject malicious code in a comment by having an href without quotes:
`<a href=javascript:alert('XSS')>0‎‎:‎30</a>`
<img width="360" height="103" alt="image" src="https://github.com/user-attachments/assets/9af5918b-8f47-41c1-ad29-49e76aa7f8ab" />
<img width="490" height="68" alt="image" src="https://github.com/user-attachments/assets/9c9aef0e-eba3-49a7-80e2-5bc6fa6d4458" />
<img width="434" height="134" alt="image" src="https://github.com/user-attachments/assets/bfb16b71-1f4b-4edc-9204-c30a619404d9" />

My proposed solution to this is relatively straight forward:

- In `formatCommentRuns`, you already are extracting the runs to seperate text, links, and timestamps
- We obviously want to make sure to properly properly links and timestamps so instead of just sanitizing the full YCS comment html, we just need to sanitize the actual comment text. This is because we know that the YCS link and timestamp html we create are safe.
- I used the encode function from the html-libraries library you are already using to convert the text into literal text. This essentially sanitizes the text in the comments (to be extra safe we could use an actual html sanitization library like DOMPurify). 
- This also solves the issue of comments that have html code being displayed incorrectly in YCS 

In the below example, you can see an image icon, now with my proposed changes, the img tag comment is safely shown.
<img width="843" height="254" alt="image" src="https://github.com/user-attachments/assets/1777c07d-a568-49c7-9b1c-ecc36ddcf462" />
<img width="844" height="271" alt="image" src="https://github.com/user-attachments/assets/313024e4-b5c9-4994-a7ab-cda4c4ded353" />

This will make searching for comments about web dev easier for users. I have included some extra test cases and commented out two since they don't work properly with the new implementation. 

I'm not 100% sure of what `fallbackText` is supposed to be so I just sanitized/encoded it just to be safe. I also fixed an issue where links would be hyperlinked to youtube.comhttps://google.com instead of the actual url by changing the canoncialURL to just be the url.

Feel free to make any edits or comment anything notable while testing, I haven't really tested livestream chats or community posts that much.

ChrisLolz committed:
Fix potential XSS attack and comment formatting

ChrisLolz committed:
fix youtube.com being appended to canonicalurl

ChrisLolz committed:
remove debug

pc035860 commented:
Thanks for catching this! While looking into it, I noticed the same XSS pattern also shows up in a couple of other places — chat replay (`formatChatRuns` in `chat/utils.ts`) and the YouTube Data API transform (`formatRunsToText` in `youtubeDataApi/transform.ts`).

I’ll add another commit to apply the same approach you used here (using `encode()` for text and `safeUrl()` for URLs) to those paths as well. I’ll also add `safeUrl()` to the comment pipeline for navigation endpoint URLs, and encode the emoji alt/title attributes.

I’ll add some test cases for these changes too.

pc035860 committed:
fix(xss): encode user content in chat, comments, and Data API

Apply html-entities encoding and safeUrl filtering across all user content rendering:

- chat/utils.ts: Encode text content, emoji alt/title, and filter dangerous URL protocols
- comments/pipeline.ts: Encode link text, emoji attributes, and sanitize navigation URLs
- youtubeDataApi/transform.ts: Encode rendered text content and filter URLs

Add comprehensive XSS security tests:
- formatChatRuns: script injection, HTML injection, javascript: URLs, emoji XSS
- formatCommentRuns: link text encoding, emoji alt encoding, URL filtering
- safeUrl: protocol blocking tests (javascript:, data:, vbscript:, file:, about:)
- YouTube Data API transform: script/HTML stripping, URL sanitization

pc035860 committed:
test(viewModels): remove obsolete commented tests

- Remove 63 lines of commented-out test code that was no longer relevant
- Remove unused import (buildChatMessageViewModels, formatCommentRuns)

pc035860 commented:
I’ll also remove a couple of escape / sanitize helpers that aren’t used anymore (`esc` and `sanitizeHTML`).

pc035860 committed:
refactor: remove unused esc and sanitizeHtml functions

These functions were replaced by html-entities encode() in the XSS security fix.

- formatting.ts: Remove esc() and sanitizeHtml() functions
- viewModels.ts: Remove duplicate sanitizeHtml() function
- formatting.test.ts: Remove corresponding test cases

pc035860 committed:
docs: document XSS prevention strategy using html-entities

Add HTML Encoding Strategy section to CLAUDE.md explaining:
- YouTube API returns plain text (not HTML entities)
- JSON Unicode escapes become literal characters after parsing
- html-entities encode() converts to safe HTML entities
- Key files involved in encoding pipeline
- Design decision rationale for direct encode() approach

pc035860 committed:
fix(xss): sanitize image src URLs in emoji and attachment rendering

Apply safeUrl() to img src attributes in chat and comment formatters
to block javascript:, data:, and other dangerous URL protocols.

- chat/utils.ts: Sanitize emoji and attachment image URLs
- comments/pipeline.ts: Sanitize emoji and attachment image URLs
- Add XSS tests for img src URL filtering
- Add double-encoding prevention tests for escapeHtml
